src/http/server.cr

```diff
@@ -95,90 +100,107 @@ class HTTP::Server
   {% end %}
 
   @wants_close = false
+  @sockets = [] of Socket::Server
 
-  def self.new(port, &handler : Context ->)
-    new("127.0.0.1", port, &handler)
+  # Creates a new HTTP server with the given block as handler.
+  def self.new(&handler : HTTP::Handler::Proc) : self
+    new(handler)
   end
 
-  def self.new(port, handlers : Array(HTTP::Handler), &handler : Context ->)
-    new("127.0.0.1", port, handlers, &handler)
+  # Creates a new HTTP server with a handler chain constructed from the *handlers*
+  # array and the given block.
+  def self.new(handlers : Array(HTTP::Handler), &handler : HTTP::Handler::Proc) : self
+    new(HTTP::Server.build_middleware(handlers, handler))
   end
 
-  def self.new(port, handlers : Array(HTTP::Handler))
-    new("127.0.0.1", port, handlers)
-  end
-
-  def self.new(port, handler)
-    new("127.0.0.1", port, handler)
-  end
-
-  def initialize(@host : String, @port : Int32, &handler : Context ->)
-    @processor = RequestProcessor.new(handler)
+  # Creates a new HTTP server with the *handlers* array as handler chain.
+  def self.new(handlers : Array(HTTP::Handler)) : self
+    new(HTTP::Server.build_middleware(handlers))
   end
 
-  def initialize(@host : String, @port : Int32, handlers : Array(HTTP::Handler), &handler : Context ->)
-    handler = HTTP::Server.build_middleware handlers, handler
+  # Creates a new HTTP server with the given *handler*.
+  def initialize(handler : HTTP::Handler | HTTP::Handler::Proc)
     @processor = RequestProcessor.new(handler)
   end
 
-  def initialize(@host : String, @port : Int32, handlers : Array(HTTP::Handler))
-    handler = HTTP::Server.build_middleware handlers
-    @processor = RequestProcessor.new(handler)
+  # Creates a `TCPServer` and adds it as a socket, returning the local address
+  # and port the server listens on.
+  #
+  # If *port* is `0`, a random, free port will be chosen.
```

Comment:
If you're going to have `bind_unused_port` then don't advertise this. Consider even disallowing it. It's platform-specific.